app/rankers/algorithms.py

```diff
@@ -246,15 +248,41 @@ def user_impression_filter(recs: RecommendationListType, user_impressed_list: Li
 def rank_by_preferred_topics(
         recs: CorpusRecommendationListType,
         preferred_topics: List[TopicModel],
+        count: int
 ) -> CorpusRecommendationListType:
     """
-    Sorts recommendations in-order such that recs with a preferred topic rank above recs with a non-preferred topic.
-    :param recs:
+    Recommends items with preferred topics, rotating the topics. Items for each topic are ordered randomly.
+    In case there's not enough items for the preferred topics, it adds the rest of the candidates in the provided order.
+
+    :param recs: candidates
     :param preferred_topics: List of topics that the user has expressed a preference for.
-    :return: ordered list of recommendations with preferred topics ranking higher.
+    :param count: a number of items to recommend
+    :return: ordered list of recommendations with preferred topics.
     """
-    preferred_corpus_topic_ids = {topic.corpus_topic_id for topic in preferred_topics}
-    return sorted(recs, key=lambda rec: rec.topic in preferred_corpus_topic_ids, reverse=True)
+    topic_ids = {topic.corpus_topic_id for topic in preferred_topics}
+    recs_by_topic = defaultdict(list)
+    for rec in recs:
+        if rec.topic in topic_ids:
+            recs_by_topic[rec.topic].append(rec)
+
+    res = []
+    recommended = set()
+    while len(res) < count and len(recs_by_topic) > 0:
+        for topic in topic_ids:
+            if topic not in recs_by_topic:
```

Comment:
Optional:
```suggestion
            if not recs_by_topic.get(topic):
```
This removes the need for a `del` statement. I initially thought that `pop` could raise an exception when `topic_recs` was empty, until I realized you handled this with the `del` statement.